docs/plutus-ir/plutus-ir.md

```diff
@@ -0,0 +1,193 @@
+we don't know how to compile those yet. I think we will eventually be able to cover all
+types defined with this syntax, but until then, validity of Plutus IR programs will be somewhat
+determined simply by *whatever we are able to compile*.
+
+### Typechecking
+
+Similarly, while we will define a typing judgement for Plutus IR, we will probably not
+write a typechecker for it, and rather will compile it to Plutus Core before typechecking.
+
+### Value restriction
+
+Since type-lets are going to be desugared as type abstractions, a type-let whose body is not
+a value will fail the Plutus Core value restriction, and there is no attempt to prevent this
+in the Plutus IR syntax. It will simply compile into an invalid Plutus Core program.
+
+## Typing
+
+Typing rules are largely the same as Plutus Core.
+
+(I should write out some real rules for this, but everything being variadic makes it a bit noisy.)
+
+Let bindings are straightforward: they introduce names into the context with their declared
+type or kind. For recursive let bindings we typecheck each binding under the assumption that
+all the bindings have their declared type or kind.
+
+We do need to state something more complex about datatype bindings: they introduce a
+number of names with various types or kinds, with the correctness conditions being:
+- The type variable declarations are well-kinded.
+- The application of the datatype to its type variables is well-kinded and of kind `(type)`.
+- The constructor argument types are all well-kinded and of type `(type)`.
+
+## Design notes
+
+### Why do we need variadic syntax?
+
+`letrec` is inherently about binding an arbitrary number of mutually recursive bindings, so
+we need to be able to write that.
+
+### Why is `let` variadic?
+
+`letrec` obviously needs to be variadic, since the whole point is to have multiple bindings
+that reference each other.
+
+`let` doesn't *need* to be variadic, but it's symmetrical, convenient, and no harder to support.
+
+### Why is there a special datatype binding?
+
+Datatype bindings are weird in that they bind multiple names, which are part of the binding
+(the type, the constructors, the destructor). So we can't handle it as a normal "type binding" since
+it also declares other things.
+
+### Why are variable declarations reified in the syntax?
+
+It's very helpful to have a syntax form for variable declarations, since the syntax is too
+homogeneous when everything is just names. For example, in the datatype declaration we have
+a whole series of type variable declarations. I think we could just about parse the version
+with them all flattened, but it's much clearer when they're reified.
+
+I think this is quite useful and arguably worth introducing in the Plutus Core syntax too.
+
+### Why does `datatype` bind its type variables?
+
+I considered having the constructors of the datatype have an explicit type instead of having arguments, e.g.
+`(constr Cons (all a (type) (fun a (fun [List a] [List a]))))`.
```

Comment:
I agree with @effectfully, I think, it would be nicer to go with GADT syntax. (The Haskell data syntax is really just a compromise so constructor declarations read like the corresponding pattern matches. As Plutus IR is not for human  consumption, I'd go with GADT syntax instead.)

Reply:
I think we should have Haskell-style rather that GADT-style data declarations. Using GADT-style makes sense when one has the full power of GADTs, but we don't. GADT-style declarations in this context are misleading, and just make life needlessly a wee bit more difficult for (a) those generating Plutus IR from Haskell data declarations (or similar) and (b) whoever implements Plutus IR.

Reply:
I honestly don't mind which we go with. I'm probably going to leave this as it is for now, and we can discuss it again later (since one of the nice points of Plutus IR is that we can change it easily!).

I agree that GADT style declarations suggest more generality than we have, but I don't think this is *too* much of a problem. On the other hand, the GADT syntax is actually simpler, because it makes the declarations of the constructors much more like normal variable declarations with a name and a type.